### PR TITLE
fix JSON schema issues

### PR DESCRIPTION
**Pull request by BlueAndi**
#760 An enumeration can be empty. In this case specify only the type "string", otherwise a empty array would be generated and causes a error.

### CODE REVIEW

**Comment by timja** [comment]
Would be nice to have a test with a fixture showing what's expected

**Comment by casz** [comment]
I wouldn't know how to test jelly 😄

**Comment by timja** [comment]
This renders JSON so should be simpler,
I think normally HTMLUnit / Selenium is used

**Comment by BlueAndi** [comment]
> Would be nice to have a test with a fixture showing what's expected

Is there already a place for jelly tests, also with an example? I am not so familiar with the structure right now.

**Comment by casz** [comment]
I would prefer we moved the output from jelly to java code... making it more testable.

**Comment by BlueAndi** [comment]
> I would prefer we moved the output from jelly to java code...

Just for clarification, I am an embedded developer. :grin:

**Comment by casz** [comment]
Perhaps we should consider using an actual schema generator?
https://github.com/FasterXML/jackson-module-jsonSchema

**Comment by BlueAndi** [comment]
> Perhaps we should consider using an actual schema generator?

A question for the roadmap!? How to deal with the current situation ... so jelly ... test script or not ... ?

**Comment by casz** [comment]
I'd still vote for java code with tests.
But if your current changes fixes all out standing issues perhaps we can merge those first and focus on converting it to java code with tests in a separate PR.

File: plugin/src/main/resources/io/jenkins/plugins/casc/DescribableAttribute/schema.jelly [hunk 1]
```diff
@@ -9,7 +9,7 @@
       <j:arg value="${it.type}"/>
     </j:invokeStatic>
       {
-        "${symbol}": { "type": "#/definitions/${d.klass.toJavaClass().name}" }
+        "${symbol}": { "$ref": "#/definitions/${d.klass.toJavaClass().name}" }
```
Comment (casz) [inline plugin/src/main/resources/io/jenkins/plugins/casc/DescribableAttribute/schema.jelly]: What about here?

File: plugin/src/main/resources/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator/schema.jelly [hunk 2]
```diff
@@ -6,7 +6,7 @@
   <j:forEach items="${it.implementors.entrySet()}" var="impl" varStatus="st">
     {
       "properties" : {
-        "${impl.key}" : { "type" : "#/definitions/${impl.value}" }
+        "${impl.key}" : { "$ref" : "#/definitions/${impl.value}" }
```
Comment (casz) [inline plugin/src/main/resources/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator/schema.jelly]: What about here?

**Comment by BlueAndi** [comment]
I will test the changes by using the built plugin out of this pull request.

**Comment by BlueAndi** [comment]
@casz Respect ... you are faster than light.

**Comment by casz** [comment]
:running_man:

File: plugin/src/main/resources/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator/schema.jelly (hunk 2 above)
Comment (BlueAndi) [reply plugin/src/main/resources/io/jenkins/plugins/casc/impl/configurators/HeteroDescribableConfigurator/schema.jelly]: Fixed in the following commit.

File: plugin/src/main/resources/io/jenkins/plugins/casc/DescribableAttribute/schema.jelly (hunk 1 above)
Comment (BlueAndi) [reply plugin/src/main/resources/io/jenkins/plugins/casc/DescribableAttribute/schema.jelly]: Fixed in the following commit.

**Comment by BlueAndi** [comment]
Number of schema errors decreased from over 160 to 11 now.
Anyway still 11 open.

**Comment by BlueAndi** [comment]
@casc All errors fixed, except the duplicated object keys, which may be fixed with #674 . So this PR is finished from my side.

**Pull request merged**